floodsub.go

```diff
@@ -324,6 +359,94 @@ func msgID(pmsg *pb.Message) string {
 	return string(pmsg.GetFrom()) + string(pmsg.GetSeqno())
 }
 
+// pushMsg pushes a message to a number of subscriptions, performing validation
+// as necessary
+func (p *PubSub) pushMsg(subs []*Subscription, src peer.ID, msg *Message) {
+	// we perform validation if _any_ of the subscriptions has a validator
+	// because the message is sent once for all topics
+	needval := false
+	for _, sub := range subs {
+		if sub.validate != nil {
+			needval = true
+			break
+		}
+	}
+
+	if needval {
+		// validation is asynchronous and globally throttled with the throttleValidate semaphore
+		select {
+		case p.validateThrottle <- struct{}{}:
+			go func() {
+				p.validate(subs, src, msg)
+				<-p.validateThrottle
+			}()
+		default:
+			log.Warningf("message validation throttled; dropping message from %s", src)
+		}
+		return
+	}
+
+	sreq := &sendReq{from: src, msg: msg}
+	select {
+	case p.sendMsg <- sreq:
+	default:
+		go func() {
```

Comment:
I'm not a fan of this pattern (too easy to open a bunch of go routines). Could we just call `maybePublishMessage` directly and rename the `sendMsg` channel name to `validatedMsgs` or something?

Reply:
I added a buffer to sendMsg so that we should almost never overflow.
But you are right, we can just call `maybePublishMessage`!
I'll keep the `sendMsg` channel named as it is, because it indicates a message that is ready to send.

Reply:
I skipped the dance, and called `maybePublishMessage` directly .